### PR TITLE
planner,executor: fix cached table query with filter condition

### LINKED ISSUE #32422: Filter condition is missing caused wrong result in cached table scanning
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
mysql> create table t (id int, c int);
Query OK, 0 rows affected (0.00 sec)

mysql> insert into t values (3,3), (4,4), (5,5);
Query OK, 3 rows affected (0.00 sec)
Records: 3  Duplicates: 0  Warnings: 0

mysql> alter table t cache;
Query OK, 1 row affected (0.01 sec)
```

### 2. What did you expect to see? (Required)

```
mysql>  select id+1, c from t where c = 4;
+------+------+
| id+1 | c    |
+------+------+
|    5 |    4 |
+------+------+
1 row in set (0.00 sec)
```

### 3. What did you see instead (Required)

```
// Before the data is cached, the result is correct....
mysql>  select id+1, c from t where c = 4;
+------+------+
| id+1 | c    |
+------+------+
|    5 |    4 |
+------+------+
1 row in set (0.00 sec)

// Execute it several times, after the data is cached, the result become wrong...
// The filter condition c = 4 doesn't take effect.
mysql>  select id+1, c from t where c = 4;
+------+------+
| id+1 | c    |
+------+------+
|    4 |    3 |
|    5 |    4 |
|    6 |    5 |
+------+------+
3 rows in set (0.00 sec)
```

### 4. What is your TiDB version? (Required)

master db3e69eb68de9acfcebd69909e7309caed64914c

The bug is introduced by https://github.com/pingcap/tidb/pull/32139

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve?

Issue Number: close #32422

Problem Summary:

This bug was introduced by https://github.com/pingcap/tidb/pull/32139
In the initial implementation, I generate UnionScan plan when reading from cache, and no UnionScan when the cache is bypassed or unavailable.
But since the shape of the plan is different, the prepared plan cache feature is not applicable.

Then, https://github.com/pingcap/tidb/pull/32139 try to fix the prepare plan cache issue by defer the UnionScanExec generation to the executor build phase. It did make plan cache possible, but introduce bug #32422

Handling the filter condition is not a trivial work if we continue generating UnionScanExec in the executor build phase:
- In the logical plan phase, by the predication push down optimize rule, the condition is seperated. 
Some are pushed down to TableReader and some remains, cause it's difficult to rebuild the filter condition for UnionScan in the executor build phase.
- And also the existence of generated column make the situation from bad to worse.

### What is changed and how it works?

So I change the implementation the third time.
This time, I will always keep the UnionScan in the planner phase for a cached table, and whether the UnionScan is really used is decided by whether the cached data is available.

In this way, prepared plan cache is applicable, since the shape of plan is unchanged (although maybe less efficient in some minor cases), and also the filter condition bug is fixed.

### Check List

Tests 

- [X] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: executor/builder.go [hunk 1]
```diff
@@ -1149,13 +1096,37 @@ func (b *executorBuilder) buildUnionScanFromReader(reader Executor, v *plannerco
...
+	}
+
 	switch x := reader.(type) {
 	case *TableReaderExecutor:
 		us.desc = x.desc
 		us.conditions, us.conditionsWithVirCol = plannercore.SplitSelCondsWithVirtualColumn(v.Conditions)
 		us.columns = x.columns
 		us.table = x.table
 		us.virtualColumnIndex = x.virtualColumnIndex
+		tbl := x.Table()
```
Comment (lcwangchao) [inline executor/builder.go:1113]: The additional codes for each child Executor are the same. It's better to move them to a method and reuse it.

Comment (tiancaiamao) [reply executor/builder.go]: Done

**Review by lcwangchao — APPROVED** [review]

**Review by tangenta — APPROVED** [review]

**Pull request merged**